include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+// ensuring it's included so callers don't have to include <string> just for network.h
+#if (WIL_USE_STL && defined(WIL_ENABLE_EXCEPTIONS)) || defined(WIL_DOXYGEN)
+#include <string>
+#endif
+
+namespace wil
+{
+// Functions and classes that support networking operations and structures
+namespace network
+{
+    // A type that calls WSACleanup on destruction (or reset()).
+    // WSAStartup must be called for the lifetime of all Winsock APIs (synchronous and asynchronous)
+    // WSACleanup will unload the full Winsock catalog - all the libraries - with the final reference
+    // which can lead to crashes if socket APIs are still being used after the final WSACleanup is called
+    using unique_wsacleanup_call = ::wil::unique_call<decltype(&::WSACleanup), ::WSACleanup>;
+
+    // Calls WSAStartup; returns an RAII object that reverts, the RAII object will resolve to bool 'false' if failed
+    WI_NODISCARD inline ::wil::network::unique_wsacleanup_call WSAStartup_nothrow() WI_NOEXCEPT
+    {
+        WSADATA unused_data{};
+        const auto error{::WSAStartup(WINSOCK_VERSION, &unused_data)};
+        LOG_IF_WIN32_ERROR(error);
+
+        ::wil::network::unique_wsacleanup_call return_cleanup{};
+        if (error != 0)
+        {
+            // internally set m_call to false
+            // so the caller can check the return object against its operator bool
+            // to determine if WSAStartup succeeded
+            return_cleanup.release();
+        }
+        return return_cleanup;
+    }
+
+    // Calls WSAStartup and fail-fasts on error; returns an RAII object that reverts
+    WI_NODISCARD inline ::wil::network::unique_wsacleanup_call WSAStartup_failfast() WI_NOEXCEPT
+    {
+        WSADATA unused_data{};
+        FAIL_FAST_IF_WIN32_ERROR(::WSAStartup(WINSOCK_VERSION, &unused_data));
+        return {};
+    }
+
+#if defined(WIL_ENABLE_EXCEPTIONS)
+    // Calls WSAStartup and throws on error; returns an RAII object that reverts
+    WI_NODISCARD inline ::wil::network::unique_wsacleanup_call WSAStartup()
+    {
+        WSADATA unused_data{};
+        THROW_IF_WIN32_ERROR(::WSAStartup(WINSOCK_VERSION, &unused_data));
+        return {};
+    }
+#endif
+
+    //
+    // utility functions to compare inaddr types
+    //
+    [[nodiscard]] inline bool equals(const ::in_addr& lhs, const ::in_addr& rhs) WI_NOEXCEPT
+    {
+        return lhs.s_addr == rhs.s_addr;
+    }
+
+    [[nodiscard]] inline bool not_equals(const ::in_addr& lhs, const ::in_addr& rhs) WI_NOEXCEPT
+    {
+        return !::wil::network::equals(lhs, rhs);
+    }
+
+    [[nodiscard]] inline bool equals(const ::in6_addr& lhs, const ::in6_addr& rhs) WI_NOEXCEPT
+    {
+        return 0 == ::memcmp(&lhs, &rhs, sizeof(::in6_addr));
+    }
+
+    [[nodiscard]] inline bool not_equals(const ::in6_addr& lhs, const ::in6_addr& rhs) WI_NOEXCEPT
+    {
+        return !::wil::network::equals(lhs, rhs);
+    }

```

Comment:
Should there also be `operator==`/`operator!=` free functions? Why `equals` and `not_equals`?

Reply:
I didn't want to overload that operator on existing structures (e.g. surprising existing code that might be using == and != with the structures today, but I just changed their behavior if/when they adding a using namespace in their cpp). so I was following something like what winrt.h does.

(I'm trying to remember the other reason I did this)

does this seem completely wrong to use these verbs as opposed to overloading those operators?

Reply:
hmm. I just played with this. another reason is that in order for operator== to be chosen, someone would needed to have declared using wil::network. which I don't want to require or even recommend. (and seems inconsistent with the other wil headers).